analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs

```diff
@@ -112,6 +112,20 @@ public ProgramState ResetOperations()
         public ProgramState RemoveSymbols(Func<ISymbol, bool> remove) =>
             this with { SymbolValue = SymbolValue.Where(kv => PreservedSymbols.Contains(kv.Key) || !remove(kv.Key)).ToImmutableDictionary() };
 
+        public ProgramState ResetFieldConstraints()
+        {
+            var state = this;
+            foreach (var kvp in SymbolValue)
+            {
+                if (kvp is { Key: IFieldSymbol fieldSymbol, Value: { } symbolValue }
+                    && symbolValue.AllConstraints().Where(x => !x.PreserveOnFieldReset(fieldSymbol)).ToArray() is { Length: > 0 } resetConstraints)
+                {
+                    state = state.SetSymbolValue(fieldSymbol, symbolValue.WithoutConstraint(resetConstraints));
+                }
+            }
```

Comment:
We don't need to rename the stuff, so it can be simplified quite a lot.
```suggestion
            foreach (var kvp in SymbolValue.Where(x => x.Key is IFieldSymbol)
            {
                if (kvp.Value.AllConstraints().Where(x => !x.PreserveOnFieldReset).ToArray() is { Length: > 0 } resetConstraints)
                {
                    state = state.SetSymbolValue(kvp.Key, kvp.Value.WithoutConstraint(resetConstraints));
                }
            }
```

Reply:
I'm not 100% convicned that the `WithoutConstraint` with array is worth the trouble. There won't be many constraints. We can go just
```
            foreach (var kvp in SymbolValue.Where(x => x.Key is IFieldSymbol)
            {
                foreach(var constraint in kvp.Value.AllConstraints().Where(x => !x.PreserveOnFieldReset))
                {
                    state = state.SetSymbolValue(kvp.Key, kvp.Value.WithoutConstraint(constraint));
                }
            }
```

Reply:
This overload uses `RemoveRange` which saves some allocations sometimes. This was the main driver for the change despite being a bit more readable.

Reply:
We will never remove more than 5 constraints (likely 1 or 2). I would not complicate the system with it